Wallet passive startup

ryanofsky opened the pull request:
Move wallet startup code closer to a simple model where the wallet attaches to the chain with a single chain.handleNotifications() call, and just starts passively receiving blocks and mempool notifications from the last update point, instead having to actively rescan blocks and request a mempool snapshot, and deal with the tip changing, and deal with early or stale notifications.

Also, stop locking the cs_wallet mutex and registering for validationinterface notifications before the rescan. This was new behavior since 6a72f26968cf931c985d8d4797b6264274cabd06 https://github.com/bitcoin/bitcoin/pull/16426 and is not ideal because it stops other wallets and rpcs and the gui from receiving new notifications until after the scan completes.

This change is a half-step towards implementing multiwallet scans (https://github.com/bitcoin/bitcoin/issues/11756), since it provides needed locator and birthday timestamp information to the Chain interface, and it rationalizes locking and event ordering in the startup code. The second half of implementing parallel rescans requires moving the ScanForWalletTransactions implementation (which this PR does not touch) from the wallet to the node.

promag commented:
How will `snapshot_fn` work with IPC?

ryanofsky commented:
re: https://github.com/bitcoin/bitcoin/pull/15719#issuecomment-478765890

> How will snapshot_fn work with IPC?

No different than other callback arguments. In #10102, `std::function` arguments are passed as objects with a single `call()` method:

https://github.com/ryanofsky/bitcoin/blob/3440513a45e94a5f1f0c67ab7409439afbbef673/src/interfaces/capnp/messages.capnp#L144-L154

https://github.com/ryanofsky/bitcoin/blob/3440513a45e94a5f1f0c67ab7409439afbbef673/src/interfaces/capnp/messages.capnp#L236-L239

practicalswift commented on src/interfaces/chain.cpp:
```diff
@@ -358,22 +359,38 @@ class ChainImpl : public Chain
     {
-        return MakeUnique<NotificationsHandlerImpl>(*this, notifications);
+        AssertLockNotHeld(::cs_main);
+        AssertLockNotHeld(::mempool.cs);
```
Could use `LOCKS_EXCLUDED` annotations for those?

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/15719#discussion_r271169011

> Could use LOCKS_EXCLUDED annotations for those?

Added annotations. Effect is kind of limited since this is an override and these mutexes aren't accessible where this is called in the wallet, but it still seems better to have these annotations than not.

Longer term, after more changes like this and #15632, which reduce Chain::Lock usage, wallet code should stop holding these locks altogether.

promag commented:
@ryanofsky ok thanks, will look into that.

promag commented on src/interfaces/chain.cpp:
```diff
@@ -358,22 +359,38 @@ class ChainImpl : public Chain
+        std::vector<CTransactionRef> snapshot;
+        std::packaged_task<std::unique_ptr<Handler>()> task([&] {
+            if (snapshot_fn) snapshot_fn(std::move(snapshot));
+            return MakeUnique<NotificationsHandlerImpl>(*this, notifications);
```
Mempool can change after the snapshot but before registering the interface and so those changes will be missed.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/15719#discussion_r271292127

> Mempool can change after the snapshot but before registering the interface and so those changes will be missed.

This isn't true, and the comment below, "Hold locks while scheduling the task so notifications about added and removed transactions after the snapshot arrive after the snapshot" is specifically referring to this case.

Let me know if I should make it more clear, but as long as the locks are held, transactions can't be added / removed by other threads, and as long as CallFunctionInValidationInterfaceQueue is called before transactions are added/removed, the handler will be registered before the queued notifications signal it.

promag replied:
But the queue can be non empty when CallFunctionInValidationInterfaceQueue is called?

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/15719#discussion_r271375746

> But the queue can be non empty when CallFunctionInValidationInterfaceQueue is called?

I think you are referring to a different problem now (problem of old notifications being received when they shouldn't be, which is different from the problem of new notifications not being received when they should be). The problem of old notifications being received when they shouldn't be is what happens without this PR in the current code:

https://github.com/bitcoin/bitcoin/blob/2c364fde423e74b4e03ebcff4582a9db7a6c4e4b/src/interfaces/chain.h#L273-L281

The problem is fixed by this PR by not constructing NotificationsHandlerImpl until after snapshot_fn is called. The comment there "to first send the snapshot before enabling subsequent notifications" is about this.

promag replied:
`AddToProcessQueue` tricked me, what really goes to that queue are notifications.

promag commented on src/interfaces/chain.cpp:
```diff
@@ -358,22 +359,39 @@ class ChainImpl : public Chain
+
+        // Snapshot mempool transactions, then schedule the task to execute.
+        // Hold locks while scheduling the task so notifications about added and
+        // removed transactions after the snapshot arrive after the snapshot.
```
also "arrive after task() is called" right?

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/15719#discussion_r271410636

> also "arrive after task() is called" right?

Yes task() is what sends the snapshot, so that's what this is referring to. Replaced "transactions after the snapshot arrive after the snapshot" with "transactions after taking the snapshot arrive after the snapshot callback in task()" to reference it and be more specific.

promag commented:
Should we assert cs_main is held in the relevant validation interface notifications?

ryanofsky commented:
re: https://github.com/bitcoin/bitcoin/pull/15719#issuecomment-479177413

> Should we assert cs_main is held in the relevant validation interface notifications?

`cs_main` can be acquired inside notification callbacks, if necessary, but it is not held when these notifications happen. This PR and #15632 disentangle notifications and locking, so locks are never held when notifications are received (right now they are inconsistently held in some cases but not others). 

Removing locking from notifications is a step toward eventually removing Chain::Lock completely, and having the wallet just asynchronously process notifications without being able to hold on to `cs_main`.

promag commented:
Sorry for not being clear. I mean that `cs_main` is locked when `GetMainSignals().TransactionAddedToMempool()` is called, and it is important that doesn't change, by adding:
```diff
@@ -152,6 +152,8 @@ void CMainSignals::UpdatedBlockTip(const CBlockIndex *pindexNew, const CBlockInd
 }

 void CMainSignals::TransactionAddedToMempool(const CTransactionRef &ptx) {
+    AssertLockHeld(::cs_main);
+    AssertLockHeld(::mempool.cs);
     m_internals->m_schedulerClient.AddToProcessQueue([ptx, this] {
         m_internals->TransactionAddedToMempool(ptx);
     });
```

jnewbery commented:
Concept ACK

> Mempool transactions are sent in a single IPC call instead of in a loop calling an IPC method repeatedly.

Do we know what this does to memory usage if there's a very large mempool (either at startup when starting one or multiple wallets, or when loading a wallet dynamically during runtime)?

ryanofsky commented:
> Do we know what this does to memory usage if there's a very large mempool (either at startup when starting one or multiple wallets, or when loading a wallet dynamically during runtime)?

With this PR by itself, a new temporary vector with one pointer per mempool transaction is created when a wallet is loaded. It's something to consider, but probably not likely to cause problems.

With this PR combined with #10102 and running `bitcoin-node` with a wallet loading in a different process, this will use significantly more memory, because serializing the vector serializes all the transactions at once in memory before sending them. This might be about as expensive as a `getrawmempool` call. If this is a problem, the `snapshot_fn` callback could be tweaked to send the transactions in batches. I guess I wouldn't want to try to optimize this prematurely, though.

Since I did write this PR "Improves performance over IPC" in the description, the increased memory usage might factor into that. But to be sure, the motivation for this PR is mainly to make notifications show up cleanly and in order on the client side.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -4213,6 +4215,17 @@ std::shared_ptr<CWallet> CWallet::CreateWalletFromFile(interfaces::Chain& chain,
+            }
+        };
+    }
+    walletInstance->m_chain_notifications_handler = chain.handleNotifications(*walletInstance, snapshot_fn);
```
@jnewbery points out that it is too early to handle mempool transactions here at this point, because the rescan hasn't happened yet. Processing a mempool transaction before the rescan could result in the `IsFromMe` check in `AddToWalletIfInvolvingMe` failing to add the transaction to the walelt when it depends on a block that hasn't been scanned yet.

So the mempool loop here needs to be moved below and will require some changes to locking, since handleNotifications in this PR can't be called with cs_main (it would deadlock).

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/15719#discussion_r274932794

> @jnewbery points out that it is too early to handle mempool transactions here at this point

This is resolved in the current version of the PR which integrates rescans and finishes recsanning before creating the mempool snapshot

jnewbery commented on src/wallet/wallet.cpp:
```diff
@@ -4213,6 +4215,17 @@ std::shared_ptr<CWallet> CWallet::CreateWalletFromFile(interfaces::Chain& chain,
     // Try to top up keypool. No-op if the wallet is locked.
     walletInstance->TopUpKeyPool();
 
+    // Register with the validation interface. Skip requesting mempool transactions if wallet is empty.
```
I think moving the sync-from-mempool function to above the initial rescan breaks the (undocumented) expectation that the wallet is notified of transactions in dependency order. In fact, this was slightly broken by #15652, but this PR makes it worse.

Prior to dynamically-loaded wallets, the wallet would always receive a transaction after it had received all of that transaction's ancestors. That wallet relies on that expectation:

- `CWallet::IsFromMe()` will only match on a transaction if its parent is already in the wallet.
- `nTimeSmart` should be monotonically increasing for transaction descendency. If transactions arrive in the wrong order then this is no longer true, and they'll be shown out-of-order in the GUI.

Since #10740, a wallet could receive mempool transactions out-of-order (since a newly loaded wallet would not be notified of transactions in the mempool, but could be notified of their descendant transactions when they arrive).  #15652 improved the situtation by notifying the wallet of all mempool transactions, but didn't sort them first, so a wallet could still be notified of the mempool txs in the wrong order. This PR potentially makes things worse by notifying the wallet of mempool txs before the blocks which could contain the mempool txs' ancestors.

I think to fix this, we need to:
- move the sync-from-mempool to after the rescan call, while cs_main is still held
- sort the mempool txs by using `mempool.infoAll()` before notifying the wallet.

We should also document that the wallet expects to see transactions in order!

ryanofsky replied:
I also wonder if there could brokenness in the case where a reorg happens during a `rescanblockchain` or `importmulti` RPC call (these rescans are different because cs_main isn't held the whole time).

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/15719#discussion_r274934195

> I think to fix this, we need to:
> 
> * move the sync-from-mempool to after the rescan call, while cs_main is still held
> * sort the mempool txs by using `mempool.infoAll()` before notifying the wallet.
> 
> We should also document that the wallet expects to see transactions in order!

Moving sync is done. Rescanning happens before syncing from mempool in the current version of this PR. Sorting mempool isn't done yet. So this will still be as broken as it has been since https://github.com/bitcoin/bitcoin/pull/15652, but not more broken. There's already a lot going on here, so I think I would want to address this in a different PR. It seems pretty easy to fix independently of this PR, but writing a test case could be a challenge

MarcoFalke commented:
@ryanofsky Are you still working on this? If no, then please close it.

ryanofsky commented:
> @ryanofsky Are you still working on this? If no, then please close it.

Thanks for the reminder. I looked into this again and I think the WIP tag still applies. I had been indecisive about involving rescan in this this change after John pointed out it was necessary, since I was unsure if it was better to incrementally change the rescan code (first synchronize rescan with notifications, then consolidate rescans across wallets https://gist.github.com/ariard/89f9bcc3a7ab9576fc6d15d251032cfa), or to just do both things in a single PR.

But now I think it should be possible to synchronize rescan with notifications in a way that doesn't change existing code too much and should also make ariard's proposed change easier.

ryanofsky committed:
refactor: Move cs_wallet to ReacceptWalletTransactions

This is just a code simplification and revert of
0440481c6bf5683eff669c789bdf6a306d99adc5 from
https://github.com/bitcoin/bitcoin/pull/15652. No behavior is changing.

ryanofsky committed:
refactor: Make CWallet:::AttachChain return scan status

Make AttachChain just responsible for syncing to the chain, moving error
handling and chain assignment to the caller before more syncing
functionality is moved in the next commit.

Also, start using the AttachChain method in tests instead of trying
attach in a more partial way using Chain::handleNotifications.

ryanofsky committed:
lib: Add FoundBlock locator support

Allow retrieving locator from Chain findBlock methods, used in upcoming commit
by the wallet to get the locator for the chain tip on startup.

ryanofsky committed:
Wallet passive startup

Move wallet startup code closer to a simple model where the wallet attaches to
the chain with a single chain.handleNotifications() call, and just starts
passively receiving blocks and mempool notifications from the last update point,
instead having to actively rescan blocks and request a mempool snapshot, and
deal with the tip changing, and deal with early or stale notifications.

Also, stop locking the cs_wallet mutex and registering for validationinterface
notifications before the rescan. This was new behavior since
6a72f26968cf931c985d8d4797b6264274cabd06
https://github.com/bitcoin/bitcoin/pull/16426 and is not ideal because it stops
other wallets and rpcs and the gui from receiving new notifications until after the
scan completes.

This change is a half-step towards implementing multiwallet parallel scans
(https://github.com/bitcoin/bitcoin/issues/11756), since it provides needed
locator and birthday timestamp information to the Chain interface, and it
rationalizes locking and event ordering in the startup code. The second half of
implementing parallel rescans requires moving the ScanForWalletTransactions
implementation (which this PR does not touch) from the wallet to the node.

ryanofsky committed:
util: Release logger lock while calling log callbacks

Prevents log callbacks from deadlocking if they call code containing log
statements. In most cases it will not make sense to call logging code from a
log hook, because it could trigger and endless chain of log messages. But if a
hook is conditioned on the content of messages, like the DebugLogHelper hook,
it can make sense and be useful.

The new functionality is used in the CreateWalletFromFile test in an upcoming
commit to be able to create transactions and blocks at specific points during
test execution and check for race conditions.

ryanofsky commented:
Rebased 3eb122b4a10024d86561492db19a881d420e34c5 -> b7716d0272125f4b4d2bce179d86d896f67869bd ([`pr/pool.9`](https://github.com/ryanofsky/bitcoin/commits/pr/pool.9) -> [`pr/pool.10`](https://github.com/ryanofsky/bitcoin/commits/pr/pool.10), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/pool.9-rebase..pr/pool.10)) after merge of #17954 and conflict with #18192

ryanofsky commented:
Rebased b7716d0272125f4b4d2bce179d86d896f67869bd -> bb7e272bd968adff75bed5038680f23b665880c5 ([`pr/pool.10`](https://github.com/ryanofsky/bitcoin/commits/pr/pool.10) -> [`pr/pool.11`](https://github.com/ryanofsky/bitcoin/commits/pr/pool.11), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/pool.10-rebase..pr/pool.11)) due to minor conflicts with #15761 and #18601

ryanofsky commented:
Rebased bb7e272bd968adff75bed5038680f23b665880c5 -> 3c50af4c76889261d19262086542b44220c74457 ([`pr/pool.11`](https://github.com/ryanofsky/bitcoin/commits/pr/pool.11) -> [`pr/pool.12`](https://github.com/ryanofsky/bitcoin/commits/pr/pool.12), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/pool.11-rebase..pr/pool.12)) due to conflict with #18759

ryanofsky commented:
Rebased 3c50af4c76889261d19262086542b44220c74457 -> 044418037a6ead2e9c38f404228ea8536369b683 ([`pr/pool.12`](https://github.com/ryanofsky/bitcoin/commits/pr/pool.12) -> [`pr/pool.13`](https://github.com/ryanofsky/bitcoin/commits/pr/pool.13), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/pool.12-rebase..pr/pool.13)) after #18727 to deal with failures in new CreateWalletFromFile test. Rewrote test to handle new startup sequence.

ryanofsky commented:
Rebased 044418037a6ead2e9c38f404228ea8536369b683 -> 26e4ee32273ab232fa3898f86289131a6379b69b ([`pr/pool.13`](https://github.com/ryanofsky/bitcoin/commits/pr/pool.13) -> [`pr/pool.14`](https://github.com/ryanofsky/bitcoin/commits/pr/pool.14), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/pool.13-rebase..pr/pool.14)) due to conflict with #16426

ryanofsky commented:
Rebased 26e4ee32273ab232fa3898f86289131a6379b69b -> 52d152f0bf2543fc6fbb80cdde7ae2249305662f ([`pr/pool.14`](https://github.com/ryanofsky/bitcoin/commits/pr/pool.14) -> [`pr/pool.15`](https://github.com/ryanofsky/bitcoin/commits/pr/pool.15), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/pool.14-rebase..pr/pool.15)) due to conflict with #18699

ryanofsky commented:
Travis error `test_framework.authproxy.JSONRPCException: non-mandatory-script-verify-flag (Signature must be zero for failed CHECK(MULTI)SIG operation) (-26)` in sendrawtransaction call in rpc_rawtransaction.py in bionic C++17 build https://travis-ci.org/github/bitcoin/bitcoin/jobs/683072285#L5042 seems unrelated

ariard commented on src/wallet/wallet.cpp:
```diff
@@ -4071,16 +4010,69 @@ bool CWallet::UpgradeWallet(int version, bilingual_str& error, std::vector<bilin
+        }
 
+        // No need to read and scan block if block was created before
+        // our wallet birthday (as adjusted for block time variability)
```
Block time variability is unclear, you mean the absence of order guarantee between `nTime`  and `nTimeFirstKey`?

ariard commented on src/interfaces/chain.h:
```diff
@@ -245,8 +228,38 @@ class Chain
+    //! Register handler for notifications. Before returning and before sending
+    //! the first notification, call ScanFn with a scan range of blocks after a
+    //! specified location and time, and then call MempoolFn with a snapshot of
+    //! transactions in the mempool before the first transactionAddedToMempool /
```
This part of comment may be made more intuitively instead of one chunk, like "Call MempoolFn with a snapshot of transaction before firing transactions back to caller".

ariard commented on src/interfaces/chain.h:
```diff
@@ -245,8 +228,38 @@ class Chain
+    //! @param[in] scan_locator  location of last block previously scanned.
+    //!                          scan_fn will be only be called for blocks after
+    //!                          this point. Can be null to scan from genesis.
+    //! @param[in] scan_time     minimum block timestamp for beginning the scan
```
May we drop either or `scan_locator` or `scan_time`, like if wallet first key is older that locator use it as a starting point, if not use wallet last block `nTime`. Do we still really need wallet locator ?

ariard commented on src/interfaces/chain.h:
```diff
@@ -245,8 +228,38 @@ class Chain
+    //!                          from this timestamp
+    //! @param[out] tip          information about chain tip at the point where
+    //!                          notifications will begin
+    virtual std::unique_ptr<Handler> handleNotifications(std::shared_ptr<Notifications> notifications,
```
I think it would be better to rename `handleNotifications`. IMO it does 3 different tasks:
* trigger rescan
* sync with mempool
* register notification

I'm fine with doing all of three in same functions rather than 3 chunks for each, but name could be something like `syncClient` ?

Also you may be interested by rescan but not mempool-sync and therefore it may be made optional.

ariard commented on src/interfaces/chain.h:
```diff
@@ -245,8 +228,38 @@ class Chain
+    //! @param[in] scan_time     minimum block timestamp for beginning the scan
+    //!                          scan_fn will only be called for blocks starting
+    //!                          from this timestamp
+    //! @param[out] tip          information about chain tip at the point where
```
You may precise tip included.

ariard commented on src/interfaces/chain.cpp:
```diff
@@ -350,9 +330,72 @@ class ChainImpl : public Chain
+
+            // Call scan_fn until it has scanned all blocks after specified
+            // location and time. Looping is necessary because new blocks may
+            // be connected during rescans.
```
Isn't taking `cs_main` lock avoid block tip increase ?

ariard reviewed (commented):
Approach ACK, thanks for moving forward with wallet initialization/rescan refactor.

Is there a way to split commit ?
* there is few-style changes around `fRescan` in `src/wallet/rcpdump.cpp`
* maybe test modifications can be moved in their own commit ?
* a wallet lock move from `import*`  in `src/rwallet/rpcdump.cpp` to `ReacceptWallet`

I'm personally fine with reviewing PR in a single chunk, but maybe other reviewers are less familiar with this part of the codebase.

ryanofsky reviewed (commented):
Updated 52d152f0bf2543fc6fbb80cdde7ae2249305662f -> 66588cfcfeb7d89f242439a97db85be44c3a2dae ([`pr/pool.15`](https://github.com/ryanofsky/bitcoin/commits/pr/pool.15) -> [`pr/pool.16`](https://github.com/ryanofsky/bitcoin/commits/pr/pool.16), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/pool.15..pr/pool.16)) splitting commits and updating some names and comments
Rebased 66588cfcfeb7d89f242439a97db85be44c3a2dae -> 78a42ce0251426051ff789e64a9bfae3200d1f83 ([`pr/pool.16`](https://github.com/ryanofsky/bitcoin/commits/pr/pool.16) -> [`pr/pool.17`](https://github.com/ryanofsky/bitcoin/commits/pr/pool.17), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/pool.16-rebase..pr/pool.17)) due to conflict with #18216

ryanofsky commented:
Updated 78a42ce0251426051ff789e64a9bfae3200d1f83 -> de030e6258251fba2159788c2bdf135f6e2369f3 ([`pr/pool.17`](https://github.com/ryanofsky/bitcoin/commits/pr/pool.17) -> [`pr/pool.18`](https://github.com/ryanofsky/bitcoin/commits/pr/pool.18), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/pool.17..pr/pool.18)) adding commit descriptions, locator test, and fixing travis macos error "comparison of integers of different signs" https://travis-ci.org/github/bitcoin/bitcoin/jobs/685957982#L3778

ariard commented:
Actually rebasing #17484, I think to avoid breaking anti-fee snipping it would be better to setup IBD state at mempool/block notification sync. Ideally calling `updateBlockTip` at any ending of block rescan. But likely I need to wait for `ScanForWalletTransactions` being moved inside the node, if I want to drop `isInitialBlockDownload`

Is this PR ready for review ?

vahidjalaliii commented:
.

در تاریخ چهارشنبه ۳ ژوئن ۲۰۲۰،‏ ۳:۰۵ DrahtBot <notifications@github.com>
نوشت:

> 🐙 This pull request conflicts with the target branch and needs rebase
> <https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#rebasing-changes>
> .
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/15719#issuecomment-637842411>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APRRJT7E4TN3UBHBPFS2U3LRUV5BNANCNFSM4HC2YJWQ>
> .
>

ryanofsky commented:
re: ariard https://github.com/bitcoin/bitcoin/pull/15719#issuecomment-638533220
> Is this PR ready for review ?

PR is complete and ready for review. Only caveat is that the unit test is triggering thread sanitizer errors. The errors look like they could be real, but I think they are errors in the test itself, not the code, and I'll look into them.

rebroad commented:
@ryanofsky I've not looked at this yet, but does this cause the GUI to become available sooner (e.g. the rescan can run after the splash screen rather than during)? I'd test this if so, or if I understand what benefits it provides - e.g. I'm not sure why we'd want parallel multiwallet scans - unless it means it would scan them all faster in total, for example.

ryanofsky commented:
Thanks for looking at this.

> @ryanofsky I've not looked at this yet, but does this cause the GUI to become available sooner (e.g. the rescan can run after the splash screen rather than during)?

Yes, this separates opening a wallet from attaching to the node, so should it be possible to interact with the wallet before the rescan finishes if GUI is changed to remove the rescan dialog or make it nonmodal.

Also if there are multiple wallets, rescanning once instead of multiple times should make loading faster in general.

> I'd test this if so, or if I understand what benefits it provides - e.g. I'm not sure why we'd want parallel multiwallet scans - unless it means it would scan them all faster in total, for example.

Updated the description to drop the word parallel. It's not really a question of parallel or serial, it's a question of scanning blocks one time for all wallets, or scanning blocks multiple times, once for each wallet.

ryanofsky commented:
Rebased 8aa64ef7f5d6e02407aa09904092e0213dcdd2c4 -> d36335f643f309790c4b26307eca622a3a9ed688 ([`pr/pool.31`](https://github.com/ryanofsky/bitcoin/commits/pr/pool.31) -> [`pr/pool.32`](https://github.com/ryanofsky/bitcoin/commits/pr/pool.32), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/pool.31-rebase..pr/pool.32)) due to conflict with #20773
